fraud/service.go

```diff
@@ -110,35 +145,112 @@ func (f *service) processIncoming(
 		f.pubsub.BlacklistPeer(from)
 		return pubsub.ValidationReject
 	}
-	log.Warnw("received fraud proof", "proofType", proof.Type(),
-		"height", proof.Height(),
-		"hash", hex.EncodeToString(extHeader.DAH.Hash()),
-		"from", from.String(),
-	)
-	msg.ValidatorData = proof
+
+	err = f.put(ctx, proof.Type(), hex.EncodeToString(proof.HeaderHash()), msg.Data)
+	if err != nil {
+		log.Error("error while adding a fraud proof to storage: ", err)
+	}
+
+	return pubsub.ValidationAccept
+}
+
+func (f *ProofService) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
-	err = put(ctx, store, string(proof.HeaderHash()), msg.Data)
-	if err != nil {
-		log.Error(err)
-	}
-	log.Warn("Shutting down services...")
-	return pubsub.ValidationAccept
+
+	return getAll(ctx, store, proofType)
 }
 
-func (f *service) Get(ctx context.Context, proofType ProofType) ([]Proof, error) {
+func (f *ProofService) put(ctx context.Context, proofType ProofType, hash string, data []byte) error {
 	f.storesLk.Lock()
 	store, ok := f.stores[proofType]
 	if !ok {
 		store = initStore(proofType, f.ds)
 		f.stores[proofType] = store
 	}
 	f.storesLk.Unlock()
+	return put(ctx, store, hash, data)
+}
 
-	return getAll(ctx, store, proofType)
+func (f *ProofService) handleFraudMessageRequest(stream network.Stream) {
+	req := &pb.FraudMessageRequest{}
+	_, err := serde.Read(stream, req)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		log.Warnf("error while handling fraud message request: %s ", err)
+		return
+	}
+
+	resp := &pb.FraudMessageResponse{}
+	errg, ctx := errgroup.WithContext(context.Background())
+	resp.Proofs = make([]*pb.ProofResponse, len(req.RequestedProofType))
+	for i, p := range req.RequestedProofType {
+		p := p
+		i := i
+		errg.Go(func() error {
+			resp.Proofs[i] = &pb.ProofResponse{Type: p}
+			proofType, err := pbToProofType(p)
+			if err != nil {
+				log.Warn(err)
+				return nil
+			}
+			proofs, err := f.Get(ctx, proofType)
+			if err != nil {
+				if errors.Is(err, datastore.ErrNotFound) {
+					return nil
+				}
+				return err
+			}
+
+			for _, proof := range proofs {
+				bin, err := proof.MarshalBinary()
+				if err != nil {
+					log.Warn(err)
+					continue
+				}
+				resp.Proofs[i].Value = bin
```

Comment:
this overwrites the value of Proofs[i] (since it's only one index per proof type) each time it loops